# Inference on Predictive and Causal Effects in High-Dimensional Nonlinear Models

## Impact of 401(k) on  Financial Wealth

As a practical illustration of the DML method, we consider estimation of the effect of 401(k) eligibility and participation
on accumulated assets. 401(k) plans are pension accounts sponsored by employers. The key problem in determining the effect of participation in 401(k) plans on accumulated assets is saver heterogeneity coupled with the fact that the decision to enroll in a 401(k) is non-random. It is generally recognized that some people have a higher preference for saving than others. It also seems likely that those individuals with high unobserved preference for saving would be most likely to choose to participate in tax-advantaged retirement savings plans and would tend to have otherwise high amounts of accumulated assets. The presence of unobserved savings preferences with these properties then implies that conventional estimates that do not account for saver heterogeneity and endogeneity of participation will be biased upward, tending to overstate the savings effects of 401(k) participation.

One can argue that eligibility for enrolling in a 401(k) plan in this data can be taken as exogenous after conditioning on a few observables of which the most important for their argument is income. The basic idea is that, at least around the time 401(k)’s initially became available, people were unlikely to be basing their employment decisions on whether an employer offered a 401(k) but would instead focus on income and other aspects of the job.

### Data

The data set can be downloaded from the github repo


In [1]:
!pip install wget

  Preparing metadata (setup.py) ... done
  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9655 sha256=4b928888ab28c9293cf8f22b325e38920e75d6a241e6e69de8eab3637dac8935
  Stored in directory: /Users/pablo/Library/Caches/pip/wheels/04/5f/3e/46cc37c5d698415694d83f607f833f83f0149e49b3af9d0f38
Successfully built wget


In [3]:
# Import relevant packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score, cross_val_predict, KFold
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LassoCV, RidgeCV, ElasticNetCV, LinearRegression, Ridge, Lasso, LogisticRegressionCV
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import GradientBoostingRegressor, GradientBoostingClassifier
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier
import patsy
import warnings
from sklearn.base import BaseEstimator, clone
import statsmodels.api as sm
import statsmodels.formula.api as smf
from IPython.display import Markdown
import wget
import os
# import seaborn as sns
warnings.simplefilter('ignore')
np.random.seed(1234)

In [4]:
file = "https://raw.githubusercontent.com/CausalAIBook/MetricsMLNotebooks/main/data/401k.csv"
data = pd.read_csv(file)

In [5]:
data.describe()

ira           a401           hval          hmort  \
count    9915.000000    9915.000000    9915.000000    9915.000000   
mean     3462.871609    3990.459304   63595.865759   30022.505093   
std      9648.027853   12825.840676   73679.374267   40056.879558   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000   48000.000000    7000.000000   
75%         0.000000     200.000000   95000.000000   50000.000000   
max    100000.000000  153000.000000  300000.000000  150000.000000   

             hequity          nifa      net_nifa           tfa       net_tfa  \
count    9915.000000  9.915000e+03  9.915000e+03  9.915000e+03  9.915000e+03   
mean    33573.360666  1.392864e+04  1.041415e+04  2.156603e+04  1.805153e+04   
std     51830.770524  5.490488e+04  5.602886e+04  6.256504e+04  6.352250e+04   
min    -40000.000000  0.000000e+00 -5.023020e+05  0.000000e+00 -5.023020e+05   
25%         0.000000  2.000000e+02 -1.350000e+03  3.750000e+02 -5.000000e+02   
50%     10000.000000  1.635000e+03  2.700000e+02  3.325000e+03  1.499000e+03   
75%     46650.000000  8.765500e+03  6.543500e+03  1.908450e+04  1.652450e+04   
max    300000.000000  1.430298e+06  1.430298e+06  1.536798e+06  1.536798e+06   

             tfa_he  ...           i3           i4           i5           i6  \
count  9.915000e+03  ...  9915.000000  9915.000000  9915.000000  9915.000000   
mean   5.162490e+04  ...     0.209178     0.172668     0.121432     0.158548   
std    9.325339e+04  ...     0.406742     0.377979     0.326645     0.365272   
min   -5.023020e+05  ...     0.000000     0.000000     0.000000     0.000000   
25%    2.000000e+02  ...     0.000000     0.000000     0.000000     0.000000   
50%    1.850000e+04  ...     0.000000     0.000000     0.000000     0.000000   
75%    6.920000e+04  ...     0.000000     0.000000     0.000000     0.000000   
max    1.687115e+06  ...     1.000000     1.000000     1.000000     1.000000   

                i7           a1           a2           a3           a4  \
count  9915.000000  9915.000000  9915.000000  9915.000000  9915.000000   
mean      0.077358     0.144024     0.209682     0.297428     0.215431   
std       0.267171     0.351132     0.407102     0.457150     0.411142   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                a5  
count  9915.000000  
mean      0.133434  
std       0.340061  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000  

[8 rows x 44 columns]

In [6]:
data.head()

ira  a401    hval  hmort  hequity   nifa  net_nifa    tfa  net_tfa  \
0     0     0   69000  60150     8850    100     -3300    100    -3300   
1     0     0   78000  20000    58000  61010     61010  61010    61010   
2  1800     0  200000  15900   184100   7549      7049   9349     8849   
3     0     0       0      0        0   2487     -6013   2487    -6013   
4     0     0  300000  90000   210000  10625     -2375  10625    -2375   

   tfa_he  ...  i3  i4  i5  i6  i7  a1  a2  a3  a4  a5  
0    5550  ...   1   0   0   0   0   0   1   0   0   0  
1  119010  ...   0   1   0   0   0   0   0   0   1   0  
2  192949  ...   0   0   0   1   0   0   0   0   1   0  
3   -6013  ...   0   0   1   0   0   1   0   0   0   0  
4  207625  ...   0   1   0   0   0   0   0   1   0   0  

[5 rows x 44 columns]

In [7]:
readme = "https://raw.githubusercontent.com/CausalAIBook/MetricsMLNotebooks/main/data/401k.md"
filename = wget.download(readme)
display(Markdown(open(filename, 'r').read()))

# Pension 401(k) data set

## Description
Data set on financial wealth and 401(k) plan participation. The source of this data set is the `hdm` package in `R`. It was created on 2/20/2023, by calling the `data(pension)` command in `R` after loading the `hdm` package and then writing the created dataframe to a csv file.

## Format
Dataframe with the following variables (amongst others):

**p401**

participation in 401(k)

**e401**

eligibility for 401(k)

**a401**

401(k) assets

**tw**

total wealth (in US $)

**tfa**

financial assets (in US $)

**net_tfa**

net financial assets (in US $)

**nifa**

non-401k financial assets (in US $)

**net_nifa**

net non-401k financial assets

**net_n401**

net non-401(k) assets (in US $)

**ira**

individual retirement account (IRA)

**inc**

income (in US $)

**age**

age

**fsize**

family size

**marr**

married

**pira**

participation in IRA

**db**

defined benefit pension

**hown**

home owner

**educ**

education (in years)

**male**

male

**twoearn**

two earners

**nohs, hs, smcol, col**

dummies for education: no high-school, high-school, some college, college

**hmort**

home mortage (in US $)

**hequity**

home equity (in US $)

**hval**

home value (in US $)

**i1-i7**

are income category indicators, 

**a1-a5**

are age category indicators

**icat** 

is the categorical variable that produces i1-i7. 

**ecat** 

is the categorical variable that generates education dummies (no high school, high school, some college, college). 

**Dum91**

is a dummy for 1991 which is equal to 1 for all observations because we are only use the 1991 SIPP. 

**tfa_he**

is net_tfa + home equity.

**zhat**

outcome of first stage prediction in instrumental variable analysis, from the academic paper that introduced this data


## Details
The sample is drawn from the 1991 Survey of Income and Program Participation (SIPP) and consists of 9,915 observations. The observational units are household reference persons aged 25-64 and spouse if present. Households are included in the sample if at least one person is employed and no one is self-employed. The data set was analysed in Chernozhukov and Hansen (2004) and Belloni et al. (2014) where further details can be found. They examine the effects of 401(k) plans on wealth using data from the Survey of Income and Program Participation using 401(k) eligibility as an instrument for 401(k) participation.

## References
V. Chernohukov, C. Hansen (2004). The impact of 401(k) participation on the wealth distribution: An instrumental quantile regression analysis. The Review of Economic and Statistics 86 (3), 735–751.

A. Belloni, V. Chernozhukov, I. Fernandez-Val, and C. Hansen (2014). Program evaluation with high-dimensional data. Working Paper.

The data consist of 9,915 observations at the household level drawn from the 1991 Survey of Income and Program Participation (SIPP). We use net financial assets (*net\_tfa*) as the outcome variable, $Y$,  in our analysis. The net financial assets are computed as the sum of IRA balances, 401(k) balances, checking accounts, saving bonds, other interest-earning accounts, other interest-earning assets, stocks, and mutual funds less non mortgage debts.

Among the $9915$ individuals, $3682$ are eligible to participate in the program. The variable *e401* indicates eligibility and *p401* indicates participation, respectively.

In [8]:
# sns.countplot(data, x='e401')
# plt.show()

Eligibility is highly associated with financial wealth:

In [9]:
# sns.displot(data=data, x='net_tfa', kind='kde', col='e401', hue='e401', fill=True)
# plt.show()

The unconditional APE of e401 is about $19559$:

In [10]:
e1 = data[data['e401'] == 1]['net_tfa']
e0 = data[data['e401'] == 0]['net_tfa']
print(f'{np.mean(e1) - np.mean(e0):.0f}')

19559


Among the $3682$ individuals that  are eligible, $2594$ decided to participate in the program. The unconditional APE of p401 is about $27372$:

In [11]:
e1 = data[data['p401'] == 1]['net_tfa']
e0 = data[data['p401'] == 0]['net_tfa']
print(f'{np.mean(e1) - np.mean(e0):.0f}')

27372


As discussed, these estimates are biased since they do not account for saver heterogeneity and endogeneity of participation.

In [12]:
y = data['net_tfa'].values
D = data['e401'].values
D2 = data['p401'].values
D3 = data['a401'].values
X = data.drop(['e401', 'p401', 'a401', 'tw', 'tfa', 'net_tfa', 'tfa_he',
               'hval', 'hmort', 'hequity',
               'nifa', 'net_nifa', 'net_n401', 'ira',
               'dum91', 'icat', 'ecat', 'zhat',
               'i1', 'i2', 'i3', 'i4', 'i5', 'i6', 'i7',
               'a1', 'a2', 'a3', 'a4', 'a5'], axis=1)
X.columns

Index(['age', 'inc', 'fsize', 'educ', 'db', 'marr', 'male', 'twoearn', 'pira',
       'nohs', 'hs', 'smcol', 'col', 'hown'],
      dtype='object')

### We define a transformer that constructs the engineered features for controls

In [13]:
!pip install formulaic

In [14]:
from sklearn.base import TransformerMixin, BaseEstimator
from formulaic import Formula

class FormulaTransformer(TransformerMixin, BaseEstimator):

    def __init__(self, formula, array=False):
        self.formula = formula
        self.array = array

    def fit(self, X, y=None):
        return self

    def transform(self, X, y=None):
        df = Formula(self.formula).get_model_matrix(X)
        if self.array:
            return df.values
        return df

In [15]:
transformer = FormulaTransformer("0 + poly(age, degree=6, raw=True) + poly(inc, degree=8, raw=True) "
                                 "+ poly(educ, degree=4, raw=True) + poly(fsize, degree=2, raw=True) "
                                 "+ male + marr + twoearn + db + pira + hown")

In [16]:
transformer.fit_transform(X).describe()

poly(age, degree=6, raw=True)[0]  poly(age, degree=6, raw=True)[1]  \
count                       9915.000000                       9915.000000   
mean                          41.060212                       1792.938981   
std                           10.344505                        899.134499   
min                           25.000000                        625.000000   
25%                           32.000000                       1024.000000   
50%                           40.000000                       1600.000000   
75%                           48.000000                       2304.000000   
max                           64.000000                       4096.000000   

       poly(age, degree=6, raw=True)[2]  poly(age, degree=6, raw=True)[3]  \
count                       9915.000000                      9.915000e+03   
mean                       82850.255371                      4.022991e+06   
std                        61725.701911                      3.939869e+06   
min                        15625.000000                      3.906250e+05   
25%                        32768.000000                      1.048576e+06   
50%                        64000.000000                      2.560000e+06   
75%                       110592.000000                      5.308416e+06   
max                       262144.000000                      1.677722e+07   

       poly(age, degree=6, raw=True)[4]  poly(age, degree=6, raw=True)[5]  \
count                      9.915000e+03                      9.915000e+03   
mean                       2.036867e+08                      1.067384e+10   
std                        2.448502e+08                      1.507229e+10   
min                        9.765625e+06                      2.441406e+08   
25%                        3.355443e+07                      1.073742e+09   
50%                        1.024000e+08                      4.096000e+09   
75%                        2.548040e+08                      1.223059e+10   
max                        1.073742e+09                      6.871948e+10   

       poly(inc, degree=8, raw=True)[0]  poly(inc, degree=8, raw=True)[1]  \
count                       9915.000000                      9.915000e+03   
mean                       37200.623197                      1.997590e+09   
std                        24774.288006                      3.026128e+09   
min                        -2652.000000                      0.000000e+00   
25%                        19413.000000                      3.768646e+08   
50%                        31476.000000                      9.907386e+08   
75%                        48583.500000                      2.360356e+09   
max                       242124.000000                      5.862403e+10   

       poly(inc, degree=8, raw=True)[2]  poly(inc, degree=8, raw=True)[3]  \
count                      9.915000e+03                      9.915000e+03   
mean                       1.437534e+14                      9.104568e+17   
std                        4.129207e+14                      3.249753e+18   
min                       -1.865179e+10                     -9.222694e+18   
25%                        7.316072e+12                      3.677463e+16   
50%                        3.118449e+13                      4.226261e+17   
75%                        1.146744e+14                      2.117477e+18   
max                        1.419428e+16                      9.222042e+18   

       ...  poly(educ, degree=4, raw=True)[2]  \
count  ...                        9915.000000   
mean   ...                        2601.853757   
std    ...                        1464.892397   
min    ...                           1.000000   
25%    ...                        1728.000000   
50%    ...                        1728.000000   
75%    ...                        4096.000000   
max    ...                        5832.000000   

       poly(educ, degree=4, raw=True)[3]  poly(fsize, degree=2, raw=True)[0]  \
count          

In [17]:
transformer = FormulaTransformer("0 + poly(age, degree=6, raw=True) + poly(inc, degree=8, raw=True) "
                                 "+ poly(educ, degree=4, raw=True) + poly(fsize, degree=2, raw=True) "
                                 "+ male + marr + twoearn + db + pira + hown", array=True)

## Estimating the ATE of 401(k) Eligibility on Net Financial Assets

We are interested in valid estimators of the average treatment effect of `e401` and `p401` on `net_tfa`. We start using ML approaches to estimate the function $g_0$ and $m_0$ in the following PLR model:

$$
\begin{align}
Y &= D\theta_0 + g_0(X) + \zeta, & E[\zeta \mid D,X]&= 0,\\
D &= m_0(X) + V, & E[V \mid X] &= 0.
\end{align}
$$

### Double ML in PLR with Cross-Fitting

We define a simple dml function that is parameterized by arbitrary ML models and returns the treatment effect and other useful quantities of the analysis

In [18]:
def dml(X, D, y, modely, modeld, *, nfolds, classifier=False):
    '''
    DML for the Partially Linear Model setting with cross-fitting

    Input
    -----
    X: the controls
    D: the treatment
    y: the outcome
    modely: the ML model for predicting the outcome y
    modeld: the ML model for predicting the treatment D
    nfolds: the number of folds in cross-fitting
    classifier: bool, whether the modeld is a classifier or a regressor

    Output
    ------
    point: the point estimate of the treatment effect of D on y
    stderr: the standard error of the treatment effect
    yhat: the cross-fitted predictions for the outcome y
    Dhat: the cross-fitted predictions for the treatment D
    resy: the outcome residuals
    resD: the treatment residuals
    epsilon: the final residual-on-residual OLS regression residual
    '''
    cv = KFold(n_splits=nfolds, shuffle=True, random_state=123) # shuffled k-folds
    yhat = cross_val_predict(modely, X, y, cv=cv, n_jobs=-1) # out-of-fold predictions for y
    # out-of-fold predictions for D
    # use predict or predict_proba dependent on classifier or regressor for D
    if classifier:
        Dhat = cross_val_predict(modeld, X, D, cv=cv, method='predict_proba', n_jobs=-1)[:, 1]
    else:
        Dhat = cross_val_predict(modeld, X, D, cv=cv, n_jobs=-1)
    # calculate outcome and treatment residuals
    resy = y - yhat
    resD = D - Dhat

    # final stage ols based point estimate and standard error
    point = np.mean(resy * resD) / np.mean(resD**2)
    epsilon = resy - point * resD
    var = np.mean(epsilon**2 * resD**2) / np.mean(resD**2)**2
    stderr = np.sqrt(var / X.shape[0])

    return point, stderr, yhat, Dhat, resy, resD, epsilon

In [19]:
def summary(point, stderr, yhat, Dhat, resy, resD, epsilon, X, D, y, *, name):
    '''
    Convenience summary function that takes the results of the DML function
    and summarizes several estimation quantities and performance metrics.
    '''
    return pd.DataFrame({'estimate': point, # point estimate
                         'stderr': stderr, # standard error
                         'lower': point - 1.96*stderr, # lower end of 95% confidence interval
                         'upper': point + 1.96*stderr, # upper end of 95% confidence interval
                         'rmse y': np.sqrt(np.mean(resy**2)), # RMSE of model that predicts outcome y
                         'rmse D': np.sqrt(np.mean(resD**2)), # RMSE of model that predicts treatment D
                         'accuracy D': np.mean(np.abs(resD) < .5), # binary classification accuracy of model for D
                         }, index=[name])

#### Double Lasso with Cross-Fitting

In [20]:
cv = KFold(n_splits=5, shuffle=True, random_state=123)
lassoy = make_pipeline(transformer, StandardScaler(), LassoCV(cv=cv))
lassod = make_pipeline(transformer, StandardScaler(), LassoCV(cv=cv))
result = dml(X, D, y, lassoy, lassod, nfolds=5)

/Users/pablo/opt/anaconda3/envs/causalML/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:664: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 4842524142.347656, tolerance: 2793965476.3442807
  model = cd_fast.enet_coordinate_descent_gram(
/Users/pablo/opt/anaconda3/envs/causalML/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:664: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 5729074177.640625, tolerance: 2793965476.3442807
  model = cd_fast.enet_coordinate_descent_gram(
/Users/pablo/opt/anaconda3/envs/causalML/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:664: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 3703135593.5898438, tolerance: 2029881536.0246825
  model = cd_fast.enet_coordinate_descent_gram(
/Users/pablo

In [21]:
table = summary(*result, X, D, y, name='double lasso')
table

estimate       stderr        lower        upper  \
double lasso  9035.120004  1295.135748  6496.653938  11573.58607   

                    rmse y    rmse D  accuracy D  
double lasso  54254.468883  0.443406    0.688553

#### Using a Penalized Logistic Regression for D

Note the default logistic regression uses an $\ell_2$ penalty. You can use the $\ell_1$ penalty as well, but computation will take longer.

In [22]:
cv = KFold(n_splits=5, shuffle=True, random_state=123)
lassoy = make_pipeline(transformer, StandardScaler(), LassoCV(cv=cv))
lgrd = make_pipeline(transformer, StandardScaler(), LogisticRegressionCV(cv=cv))
result = dml(X, D, y, lassoy, lgrd, nfolds=5, classifier=True)

/Users/pablo/opt/anaconda3/envs/causalML/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:664: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 4842524142.347656, tolerance: 2793965476.3442807
  model = cd_fast.enet_coordinate_descent_gram(
/Users/pablo/opt/anaconda3/envs/causalML/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:664: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 5729074177.640625, tolerance: 2793965476.3442807
  model = cd_fast.enet_coordinate_descent_gram(
/Users/pablo/opt/anaconda3/envs/causalML/lib/python3.9/site-packages/sklearn/linear_model/_coordinate_descent.py:664: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 3703135593.5898438, tolerance: 2029881536.0246825
  model = cd_fast.enet_coordinate_descent_gram(
/Users/pablo

In [23]:
summary(*result,  X, D, y, name='lasso/logistic')

estimate       stderr        lower         upper  \
lasso/logistic  9092.461173  1304.397656  6535.841766  11649.080579   

                      rmse y    rmse D  accuracy D  
lasso/logistic  54254.468883  0.444043    0.687847

In [24]:
table = pd.concat([table , summary(*result,  X, D, y, name='lasso/logistic')])
table

estimate       stderr        lower         upper  \
double lasso    9035.120004  1295.135748  6496.653938  11573.586070   
lasso/logistic  9092.461173  1304.397656  6535.841766  11649.080579   

                      rmse y    rmse D  accuracy D  
double lasso    54254.468883  0.443406    0.688553  
lasso/logistic  54254.468883  0.444043    0.687847

Then, we repeat this procedure for various machine learning methods.

### Random Forests

In [25]:
rfy = make_pipeline(transformer, RandomForestRegressor(n_estimators=100, min_samples_leaf=10, ccp_alpha=.001))
rfd = make_pipeline(transformer, RandomForestClassifier(n_estimators=100, min_samples_leaf=10, ccp_alpha=.001))
result = dml(X, D, y, rfy, rfd, nfolds=5, classifier=True)

In [26]:
table = pd.concat([table , summary(*result,  X, D, y, name='random forest')])
table

estimate       stderr        lower         upper  \
double lasso    9035.120004  1295.135748  6496.653938  11573.586070   
lasso/logistic  9092.461173  1304.397656  6535.841766  11649.080579   
random forest   8823.389678  1360.309746  6157.182577  11489.596779   

                      rmse y    rmse D  accuracy D  
double lasso    54254.468883  0.443406    0.688553  
lasso/logistic  54254.468883  0.444043    0.687847  
random forest   55152.962053  0.444324    0.688553

### Decision Trees

In [27]:
dtry = make_pipeline(transformer, DecisionTreeRegressor(min_samples_leaf=10, ccp_alpha=.001))
dtrd = make_pipeline(transformer, DecisionTreeClassifier(min_samples_leaf=10, ccp_alpha=.001))
result = dml(X, D, y, dtry, dtrd, nfolds=5, classifier=True)

In [28]:
table = pd.concat([table , summary(*result,  X, D, y, name='decision tree')])
table

estimate       stderr        lower         upper  \
double lasso    9035.120004  1295.135748  6496.653938  11573.586070   
lasso/logistic  9092.461173  1304.397656  6535.841766  11649.080579   
random forest   8823.389678  1360.309746  6157.182577  11489.596779   
decision tree   9237.030821  1440.550571  6413.551700  12060.509941   

                      rmse y    rmse D  accuracy D  
double lasso    54254.468883  0.443406    0.688553  
lasso/logistic  54254.468883  0.444043    0.687847  
random forest   55152.962053  0.444324    0.688553  
decision tree   59427.226868  0.446437    0.688048

### Boosted Trees

In [29]:
gbfy = make_pipeline(transformer, GradientBoostingRegressor(max_depth=2, n_iter_no_change=5))
gbfd = make_pipeline(transformer, GradientBoostingClassifier(max_depth=2, n_iter_no_change=5))
result = dml(X, D, y, gbfy, gbfd, nfolds=5, classifier=True)

In [30]:
table = pd.concat([table , summary(*result,  X, D, y, name='boosted forest')])
table

estimate       stderr        lower         upper  \
double lasso    9035.120004  1295.135748  6496.653938  11573.586070   
lasso/logistic  9092.461173  1304.397656  6535.841766  11649.080579   
random forest   8823.389678  1360.309746  6157.182577  11489.596779   
decision tree   9237.030821  1440.550571  6413.551700  12060.509941   
boosted forest  9082.510152  1336.345327  6463.273310  11701.746994   

                      rmse y    rmse D  accuracy D  
double lasso    54254.468883  0.443406    0.688553  
lasso/logistic  54254.468883  0.444043    0.687847  
random forest   55152.962053  0.444324    0.688553  
decision tree   59427.226868  0.446437    0.688048  
boosted forest  55593.735380  0.443375    0.690066

The best model with lowest RMSE in both equation is the PLR model estimated via lasso.

### AutoML

In [31]:
# !pip install flaml

In [32]:
from flaml import AutoML
flamly = make_pipeline(transformer, AutoML(time_budget=100, task='regression', early_stop=True,
                                    eval_method='cv', n_splits=3, metric='r2', verbose=3))
flamld = make_pipeline(transformer, AutoML(time_budget=100, task='classification', early_stop=True,
                                           eval_method='cv', n_splits=3, metric='r2', verbose=3))
result = dml(X, D, y, flamly, flamld, nfolds=5, classifier=True)

[flaml.automl.logger: 06-30 11:15:30] {1680} INFO - task = regression
[flaml.automl.logger: 06-30 11:15:30] {1680} INFO - task = regression
[flaml.automl.logger: 06-30 11:15:30] {1691} INFO - Evaluation method: cv
[flaml.automl.logger: 06-30 11:15:30] {1691} INFO - Evaluation method: cv
[flaml.automl.logger: 06-30 11:15:30] {1680} INFO - task = regression
[flaml.automl.logger: 06-30 11:15:30] {1691} INFO - Evaluation method: cv
[flaml.automl.logger: 06-30 11:15:30] {1789} INFO - Minimizing error metric: 1-r2
[flaml.automl.logger: 06-30 11:15:30] {1789} INFO - Minimizing error metric: 1-r2
[flaml.automl.logger: 06-30 11:15:30] {1789} INFO - Minimizing error metric: 1-r2
[flaml.automl.logger: 06-30 11:15:30] {1901} INFO - List of ML learners in AutoML Run: ['lgbm', 'rf', 'xgboost', 'extra_tree', 'xgb_limitdepth']
[flaml.automl.logger: 06-30 11:15:30] {2219} INFO - iteration 0, current learner lgbm
[flaml.automl.logger: 06-30 11:15:30] {1901} INFO - List of ML learners in AutoML Run: ['lg

KeyboardInterrupt: 

In [ ]:
table = pd.concat([table , summary(*result,  X, D, y, name='automl')])
table

# Semi-Cross-Fitting

To avoid the computational cost of performing model selection within each fold (assuming that we don't select among an exponential set of hyperparameters/models in the number of samples), it is ok to perform model selection using all the data and then perform cross-fitting with the selected model.

In [33]:
flamly = make_pipeline(transformer, AutoML(time_budget=100, task='regression', early_stop=True,
                                    eval_method='cv', n_splits=3, metric='r2', verbose=0))
flamld = make_pipeline(transformer, AutoML(time_budget=100, task='classification', early_stop=True,
                                           eval_method='cv', n_splits=3, metric='r2', verbose=0))

In [34]:
flamly.fit(X, y)
besty = make_pipeline(transformer, clone(flamly[-1].best_model_for_estimator(flamly[-1].best_estimator)))

In [35]:
flamld.fit(X, D)
bestd = make_pipeline(transformer, clone(flamld[-1].best_model_for_estimator(flamld[-1].best_estimator)))

In [36]:
result = dml(X, D, y, besty, bestd, nfolds=5, classifier=True)

In [37]:
table = pd.concat([table , summary(*result,  X, D, y, name='automl (semi-cfit)')])
table

estimate       stderr        lower         upper  \
double lasso        9035.120004  1295.135748  6496.653938  11573.586070   
lasso/logistic      9092.461173  1304.397656  6535.841766  11649.080579   
random forest       8823.389678  1360.309746  6157.182577  11489.596779   
decision tree       9237.030821  1440.550571  6413.551700  12060.509941   
boosted forest      9082.510152  1336.345327  6463.273310  11701.746994   
automl (semi-cfit)  8884.424774  1306.176253  6324.319318  11444.530230   

                          rmse y    rmse D  accuracy D  
double lasso        54254.468883  0.443406    0.688553  
lasso/logistic      54254.468883  0.444043    0.687847  
random forest       55152.962053  0.444324    0.688553  
decision tree       59427.226868  0.446437    0.688048  
boosted forest      55593.735380  0.443375    0.690066  
automl (semi-cfit)  53908.841872  0.443595    0.690973

### Semi-Crossfitting with Stacking

In [ ]:
def dml_dirty(X, D, y, modely_list, modeld_list, *,
              stacker=LinearRegression(), nfolds, classifier=False):
    '''
    DML for the Partially Linear Model setting with semi-cross-fitting

    Input
    -----
    X: the controls
    D: the treatment
    y: the outcome
    modely: the ML model for predicting the outcome y
    modeld: the ML model for predicting the treatment D
    stacker: model used to aggregate predictions of each of the base models
    nfolds: the number of folds in cross-fitting
    classifier: bool, whether the modeld is a classifier or a regressor

    Output
    ------
    point: the point estimate of the treatment effect of D on y
    stderr: the standard error of the treatment effect
    yhat: the cross-fitted predictions for the outcome y
    Dhat: the cross-fitted predictions for the treatment D
    resy: the outcome residuals
    resD: the treatment residuals
    epsilon: the final residual-on-residual OLS regression residual
    '''
    # construct out-of-fold predictions for each model
    cv = KFold(n_splits=nfolds, shuffle=True, random_state=123)
    yhats = np.array([cross_val_predict(modely, X, y, cv=cv, n_jobs=-1) for modely in modely_list]).T
    if classifier:
        Dhats = np.array([cross_val_predict(modeld, X, D, cv=cv, method='predict_proba', n_jobs=-1)[:, 1]
                         for modeld in modeld_list]).T
    else:
        Dhats = np.array([cross_val_predict(modeld, X, D, cv=cv, n_jobs=-1) for modeld in modeld_list]).T
    # calculate stacked residuals by finding optimal coefficients
    # and weigthing out-of-sample predictions by these coefficients
    yhat = stacker.fit(yhats, y).predict(yhats)
    Dhat = stacker.fit(Dhats, D).predict(Dhats)
    resy = y - yhat
    resD = D - Dhat
    # go with the stacked residuals
    point = np.mean(resy * resD) / np.mean(resD**2)
    epsilon = resy - point * resD
    var = np.mean(epsilon**2 * resD**2) / np.mean(resD**2)**2
    stderr = np.sqrt(var / X.shape[0])
    return point, stderr, yhat, Dhat, resy, resD, epsilon

In [ ]:
result = dml_dirty(X, D, y, [lassoy, rfy, dtry, gbfy], [lgrd, rfd, dtrd, gbfd],
                   nfolds=5, classifier=True)

In [ ]:
table = pd.concat([table , summary(*result, X, D, y, name='stacked (semi-cfit)')])
table

## Interactive Regression Model (IRM)

Next, we consider estimation of average treatment effects when treatment effects are fully heterogeneous:

 \begin{eqnarray}\label{eq: HetPL1}
 & Y  = g_0(D, X) + U,  &  \quad E[U \mid X, D]= 0,\\
  & D  = m_0(X) + V,  & \quad  E[V\mid X] = 0.
\end{eqnarray}

To reduce the disproportionate impact of extreme propensity score weights in the interactive model
we trim the propensity scores which are below .01 or above .99.

In [ ]:
def dr(X, D, y, modely0, modely1, modeld, *, trimming=0.01, nfolds):
    '''
    DML for the Interactive Regression Model setting (Doubly Robust Learning)
    with cross-fitting

    Input
    -----
    X: the controls
    D: the treatment
    y: the outcome
    modely0: the ML model for predicting the outcome y in the control population
    modely1: the ML model for predicting the outcome y in the treated population
    modeld: the ML model for predicting the treatment D
    trimming: threshold below which to trim propensities
    nfolds: the number of folds in cross-fitting

    Output
    ------
    point: the point estimate of the treatment effect of D on y
    stderr: the standard error of the treatment effect
    yhat: the cross-fitted predictions for the outcome y
    Dhat: the cross-fitted predictions for the outcome D
    resy: the outcome residuals
    resD: the treatment residuals
    drhat: the doubly robust quantity for each sample
    '''
    cv = KFold(n_splits=nfolds, shuffle=True, random_state=123)
    yhat0, yhat1 = np.zeros(y.shape), np.zeros(y.shape)
    # we will fit a model E[Y| D, X] by fitting a separate model for D==0
    # and a separate model for D==1.
    for train, test in cv.split(X, y):
        # train a model on training data that received treatment zero and predict on all data in test set
        yhat0[test] = clone(modely0).fit(X.iloc[train][D[train]==0], y[train][D[train]==0]).predict(X.iloc[test])
        # train a model on training data that received treatment one and predict on all data in test set
        yhat1[test] = clone(modely1).fit(X.iloc[train][D[train]==1], y[train][D[train]==1]).predict(X.iloc[test])
    # prediction for observed treatment
    yhat = yhat0 * (1 - D) + yhat1 * D
    # propensity scores
    Dhat = cross_val_predict(modeld, X, D, cv=cv, method='predict_proba', n_jobs=-1)[:, 1]
    Dhat = np.clip(Dhat, trimming, 1 - trimming)
    # doubly robust quantity for every sample
    drhat = yhat1 - yhat0 + (y - yhat) * (D/Dhat - (1 - D)/(1 - Dhat))
    point = np.mean(drhat)
    var = np.var(drhat)
    stderr = np.sqrt(var / X.shape[0])
    return point, stderr, yhat, Dhat, y - yhat, D - Dhat, drhat

**NB**: There is randomness across the random seed that potentially causes instability in our estimates. In particular, we find that the lasso/logistic specification for the IRM model exhibits great instability across different seeds. To mitigate these differences, we can take an average across seeds. In principle we would do this for all estimates, but for computation and simplicity, we cheat a little bit and do it only for the model specification we know varies significantly.

In [ ]:
cv = KFold(n_splits=5, shuffle=True, random_state=123)
lassoytest = make_pipeline(transformer, StandardScaler(), LassoCV(cv=cv))
lgrdtest = make_pipeline(transformer, StandardScaler(), LogisticRegressionCV(cv=cv))
result = dr(X, D, y, lassoytest, lassoytest, lgrdtest, nfolds=5)
seed_estimates = summary(*result,  X, D, y, name='lasso/logistic')

for i in range(9):
  cv = KFold(n_splits=5, shuffle=True, random_state=i)
  lassoytest = make_pipeline(transformer, StandardScaler(), LassoCV(cv=cv))
  lgrdtest = make_pipeline(transformer, StandardScaler(), LogisticRegressionCV(cv=cv))
  result = dr(X, D, y, lassoytest, lassoytest, lgrdtest, nfolds=5)
  seed_estimates = pd.concat([seed_estimates , summary(*result,  X, D, y, name='lasso/logistic')])

seed_estimates

As we can see, using lasso and logistic regression for $Y$ and $D$ respectively leads to estimates that vary across different seeds. Letting $\hat{\theta}_s$ denote the vector of $s=10$ estimates across the seeds, we take $$\hat{\theta} := \text{median}(\hat{\theta}_s)$$ and $$SE(\hat{\theta}) := \sqrt{\text{median}\left(SE(\hat{\theta}_s)^2 + (\hat{\theta}_s - \hat{\theta})^2\right)}$$ (Note in the standard error calculation, the operation of de-medianing $\hat{\theta}_s - \hat{\theta}$ is broadcasted.)

In [ ]:
med_theta = np.median(seed_estimates.values[:,0])
se_med = np.sqrt(np.median((seed_estimates.values[:,1])**2 + (seed_estimates.values[:,0] - med_theta)**2))
tabledr = pd.DataFrame({'estimate': med_theta,
                        'stderr': se_med,
                        'lower': med_theta - 1.96*se_med,
                        'upper': med_theta + 1.96*se_med,
                        'rmse y': np.median(seed_estimates.values[:,4]),
                        'rmse D': np.median(seed_estimates.values[:,5]),
                        'accuracy D': np.median(seed_estimates.values[:,6]),
                        }, index=['lasso/logistic'])

In [ ]:
rfy = make_pipeline(transformer, RandomForestRegressor(n_estimators=100, min_samples_leaf=10, ccp_alpha=.001))
rfd = make_pipeline(transformer, RandomForestClassifier(n_estimators=100, min_samples_leaf=10, ccp_alpha=.001))
result = dr(X, D, y, rfy, rfy, rfd, nfolds=5)

In [ ]:
tabledr = pd.concat([tabledr , summary(*result,  X, D, y, name='random forest')])
tabledr

In [ ]:
dtry = make_pipeline(transformer, DecisionTreeRegressor(min_samples_leaf=10, ccp_alpha=.001))
dtrd = make_pipeline(transformer, DecisionTreeClassifier(min_samples_leaf=10, ccp_alpha=.001))
result = dr(X, D, y, dtry, dtry, dtrd, nfolds=5)

In [ ]:
tabledr = pd.concat([tabledr , summary(*result,  X, D, y, name='decision tree')])
tabledr

In [ ]:
gbfy = make_pipeline(transformer, GradientBoostingRegressor(max_depth=2, n_iter_no_change=5))
gbfd = make_pipeline(transformer, GradientBoostingClassifier(max_depth=2, n_iter_no_change=5))
result = dr(X, D, y, gbfy, gbfy, gbfd, nfolds=5)

In [ ]:
tabledr = pd.concat([tabledr , summary(*result, X, D, y, name='boosted forest')])
tabledr

These estimates that flexibly account for confounding are substantially attenuated relative to the baseline estimate (19559) that does not account for confounding. They suggest much smaller causal effects of 401(k) eligiblity on financial asset holdings.

# Semi-Cross-Fitting

In [ ]:
from flaml import AutoML
flamly0 = make_pipeline(transformer, AutoML(time_budget=60, task='regression', early_stop=True,
                                     eval_method='cv', n_splits=3, metric='r2', verbose=0))
flamly1 = make_pipeline(transformer, AutoML(time_budget=60, task='regression', early_stop=True,
                                     eval_method='cv', n_splits=3, metric='r2', verbose=0))
flamld = make_pipeline(transformer, AutoML(time_budget=60, task='classification', early_stop=True,
                                           eval_method='cv', n_splits=3, metric='r2', verbose=0))

In [ ]:
flamly0.fit(X[D==0], y[D==0])
besty0 = make_pipeline(transformer, clone(flamly0[-1].best_model_for_estimator(flamly0[-1].best_estimator)))

In [ ]:
flamly1.fit(X[D==1], y[D==1])
besty1 = make_pipeline(transformer, clone(flamly1[-1].best_model_for_estimator(flamly1[-1].best_estimator)))

In [ ]:
flamld.fit(X, D)
bestd = make_pipeline(transformer, clone(flamld[-1].best_model_for_estimator(flamld[-1].best_estimator)))

In [ ]:
result = dr(X, D, y, besty0, besty1, bestd, nfolds=5)

In [ ]:
tabledr = pd.concat([tabledr , summary(*result,  X, D, y, name='automl (semi-cfit)')])
tabledr

In [ ]:
def dr_dirty(X, D, y, modely0_list, modely1_list, modeld_list, *,
             stacker=LinearRegression(), trimming=0.01, nfolds):
    '''
    DML for the Interactive Regression Model setting (Doubly Robust Learning)
    with cross-fitting

    Input
    -----
    X: the controls
    D: the treatment
    y: the outcome
    modely_list: list of ML models for predicting the outcome y
    modeld_list: list of ML models for predicting the treatment D
    stacker: model used to aggregate predictions of each of the base models
    trimming: threshold below which to trim propensities
    nfolds: the number of folds in cross-fitting

    Output
    ------
    point: the point estimate of the treatment effect of D on y
    stderr: the standard error of the treatment effect
    yhat: the cross-fitted predictions for the outcome y
    Dhat: the cross-fitted predictions for the outcome D
    resy: the outcome residuals
    resD: the treatment residuals
    drhat: the doubly robust quantity for each sample
    '''
    cv = KFold(n_splits=nfolds, shuffle=True, random_state=123)

    # we will fit a model E[Y| D, X] by fitting a separate model for D==0
    # and a separate model for D==1. We do that for each model type in modely_list
    yhats0, yhats1 = np.zeros((y.shape[0], len(modely0_list))), np.zeros((y.shape[0], len(modely1_list)))
    for train, test in cv.split(X, y):
        for it, modely0 in enumerate(modely0_list):
            yhats0[test, it] = clone(modely0).fit(X.iloc[train][D[train]==0], y[train][D[train]==0]).predict(X.iloc[test])
        for it, modely1 in enumerate(modely1_list):
            yhats1[test, it] = clone(modely1).fit(X.iloc[train][D[train]==1], y[train][D[train]==1]).predict(X.iloc[test])

    # calculate stacking weights for the outcome model for each population
    # and combine the outcome model predictions
    yhat0 = clone(stacker).fit(yhats0[D==0], y[D==0]).predict(yhats0)
    yhat1 = clone(stacker).fit(yhats1[D==1], y[D==1]).predict(yhats1)

    # prediction for observed treatment using the stacked model
    yhat = yhat0 * (1 - D) + yhat1 * D

    # propensity scores
    Dhats = np.array([cross_val_predict(modeld, X, D, cv=cv, method='predict_proba', n_jobs=-1)[:, 1]
                     for modeld in modeld_list]).T
    # construct coefficients on each model based on stacker
    Dhat = clone(stacker).fit(Dhats, D).predict(Dhats)
    # trim propensities
    Dhat = np.clip(Dhat, trimming, 1 - trimming)

    # doubly robust quantity for every sample
    drhat = yhat1 - yhat0 + (y - yhat) * (D/Dhat - (1 - D)/(1 - Dhat))
    point = np.mean(drhat)
    var = np.var(drhat)
    stderr = np.sqrt(var / X.shape[0])
    return point, stderr, yhat, Dhat, y - yhat, D - Dhat, drhat

In [ ]:
result = dr_dirty(X, D, y, [lassoy, rfy, dtry, gbfy], [lassoy, rfy, dtry, gbfy], [lgrd, rfd, dtrd, gbfd], nfolds=5)

In [ ]:
tabledr = pd.concat([tabledr , summary(*result,  X, D, y, name='stacked (semi-cfit)')])
tabledr

We can compare the results between the PLR model and IRM model. We find that the effect under the IRM model is typically of lower value.

In [ ]:
table

In [ ]:
tabledr

# Using the EconML Library



We are interested in valid estimators of the average treatment effect of `e401` on `net_tfa`. There exist nice packages out there that can help us do our estimation with the simple call of a function. Such packages include `EconML` (Python), `DoubleML` (Python and R), and `ddml` (Stata and R).

We run through PLR and IRM using `EconML` below to illustrate.

In [ ]:
# code verified with this version of econml
# If you want to try out the latest econml version remove the version number;
# albeit mistakes could potentially be raised due to library updates”
!pip install econml==0.14.1 #

In [ ]:
# for these libraries we will just pre-featurize the controls
W = StandardScaler().fit_transform(transformer.fit_transform(X))

In [ ]:
from econml.dml import LinearDML
cv = KFold(n_splits=5, shuffle=True, random_state=123)
ldml = LinearDML(model_y=LassoCV(cv=cv), model_t=LogisticRegressionCV(cv=cv),
                 cv=3, discrete_treatment=True, random_state=123).fit(y, D, W=W)

In [ ]:
ldml.summary()

In [ ]:
# r2scores
r2scorey = np.mean(ldml.nuisance_scores_y)
r2scored = np.mean(ldml.nuisance_scores_t)
r2scorey, r2scored

In [ ]:
from econml.dr import LinearDRLearner

# dr learner in econml fits a single regression function of Y from X, D
# using all the data
dr = LinearDRLearner(model_regression=LassoCV(cv=cv),
                     model_propensity=LogisticRegressionCV(cv=cv),
                     cv=3, min_propensity=0.01, random_state=123).fit(y, D, W=W)

In [ ]:
dr.summary(T=1)

In [ ]:
from econml.dr import LinearDRLearner
from econml.utilities import SeparateModel

# to implement the separate regression models we need use the separate model wrapper
# that splits the data based on the last covariate (in this case D) and fits a separate
# model for each group. The input to the wrapper is the model to use for each value
# of the last covariate.
dr = LinearDRLearner(model_regression=SeparateModel(LassoCV(cv=cv), LassoCV(cv=cv)),
                     model_propensity=LogisticRegressionCV(cv=cv),
                     cv=3, min_propensity=0.01, random_state=123).fit(y, D, W=W)

In [ ]:
dr.summary(T=1)

# Using the DoubleML library

We now play with `DoubleML` to illustrate its use.

In [1]:
!pip install doubleml

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 274.5/274.5 kB 9.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.3/17.3 MB 65.0 MB/s eta 0:00:00a 0:00:01


In [ ]:
from doubleml import DoubleMLData
dml_data = DoubleMLData.from_arrays(W, y, D)
print(dml_data)

In [ ]:
import doubleml as dml

dml_plr_obj = dml.DoubleMLPLR(dml_data, LassoCV(cv=cv), LogisticRegressionCV(cv=cv), n_folds=3)
print(dml_plr_obj.fit())

In [ ]:
dml_irm_obj = dml.DoubleMLIRM(dml_data, LassoCV(cv=cv), LogisticRegressionCV(cv=cv), n_folds=3)

print(dml_irm_obj.fit())